runtime/shared_common/shrinit.cpp

```diff
@@ -4923,4 +4929,147 @@ isFreeDiskSpaceLow(J9JavaVM *vm, U_64* maxsize)
+	UDATA i = 0;
+	UDATA nOptions = vm->vmArgsArray->nOptions;
+	char* key = NULL;
+	bool firstOption = true;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"-Dsun.java.launcher.pid="))
+		){
+			keyLength += strlen(actualArgs->options[i].optionString);
+			extraLen += 1;
+		}
+	}
+	if (keyLength == 0) {
+		goto done;
+	}
+	keyLength += extraLen; /* space between options and terminating null char */
+
+	key = (char*)j9mem_allocate_memory(keyLength, J9MEM_CATEGORY_VM);
+	if (NULL == key) {
+		goto done;
+	}
+	memset(key, 0, keyLength);
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"sun.java.launcher.pid"))
+		) {
+			if (firstOption) {
+				firstOption = false;
+				j9str_printf(PORTLIB, key, keyLength, "%s%s", key, option);
+			} else {
+				j9str_printf(PORTLIB, key, keyLength, "%s%s%s", key, " ", option);
+			}
+		}
+	}
+done:
+	return key;
+}
+
+/**
+ * Stores the GC hints into the shared classes cache.
+ * @param[in] vmThread  The current thread
+ * @param[in] hintsFlags Extra flags associated with the GC hints
+ * @param[in] heapSize1  The first heap size param that is to be stored into the shared cache
+ * @param[in] heapSize2  The second heap size param that is to be stored into the shared cache
+ * @param[in] forceReplace TRUE Replace the existing GC hints under the same key (the same command line arguments) if there is one already in the shared cache.
+ * 							FALSE Do not replace existing GC hints under the same key.
+ *
+ * @return  The location of the cached GC hints or null
+ */
+const U_8*
+j9shr_storeGCHints(J9VMThread* currentThread, UDATA hintsFlags, UDATA heapSize1, UDATA heapSize2, BOOLEAN forceReplace)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+	const U_8* returnVal = NULL;
+	char *key = generateGCHintsKey(vm);
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	Trc_SHR_INIT_j9shr_storeGCHints_Entry(currentThread, hintsFlags, heapSize1, heapSize2, forceReplace);
+
+	if (NULL != key) {
+		J9SharedGCHintsDataDescriptor hints = {0};
+		hints.flags = hintsFlags;
+		hints.heapSize1 = heapSize1;
+		hints.heapSize2 = heapSize2;
+		J9SharedDataDescriptor dataDescriptor = {0};
+		dataDescriptor.address = (U_8*)&hints;
+		dataDescriptor.length = sizeof(J9SharedGCHintsDataDescriptor);
+		dataDescriptor.type = J9SHR_DATA_TYPE_GC_HEAPHINT_2;
+		if (TRUE == forceReplace) {
+			dataDescriptor.flags = J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE;
+		} else {
+			dataDescriptor.flags = J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE;
+		}
+		returnVal = j9shr_storeSharedData(currentThread, key, strlen(key), &dataDescriptor);
+		j9mem_free_memory(key);
+	} else {
+		Trc_SHR_INIT_j9shr_storeGCHints_Null_Key(currentThread);
+	}
+	Trc_SHR_INIT_j9shr_storeGCHints_Exit(currentThread, returnVal);
```

Comment:
It would be better to have two exit tracepoints, one for success and another of the exception type for failure.